### PR TITLE
sys_usbd: Emulate Disney Infinity Base

### PR DESCRIPTION
**Pull request by deReeperJosh**
Having recently made the same changes for Dolphin - I am bringing the code across to Emulate the Disney Infinity Base on RPCS3. The main appeal for this is the wider variety of figures and games available to play - the Wii only has 1.0 (and it runs fairly poorly due to the game's extensive MMU access). I haven't been able to test this much, as the games don't run very well on my M1 Mac - but hoping someone interested will be able to test this out on a powerful Windows device!

There will likely be changes or RPCS3 utils that I can use instead of what I have used currently, but I have largely tried to replicate the same code styling as the skylander dialog tool.

### CODE REVIEW

**Comment by deReeperJosh** [comment]
Will need to log in to a windows machine to generate a unique id for the vcxproj - will do that when I get the chance

File: rpcs3/Emu/Io/Infinity.cpp [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+	reply_buf[3] = 0x00;
+	u8 file_block = 0;
+	if (block == 0)
+	{
+		file_block = 1;
+	}
+	else
+	{
+		file_block = block * 4;
+	}
```
Comment (Megamouse) [inline rpcs3/Emu/Io/Infinity.cpp:241]:
```suggestion
	const u8 file_block = (block == 0) ? 1 : (block * 4);
```

File: rpcs3/Emu/Io/Infinity.cpp [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+			memcpy(buf, g_infinitybase.pop_added_removed_response().data(), 0x20);
+		}
+		else
+		{
+			if (!q_queries.empty())
+			{
+				memcpy(buf, q_queries.front().data(), 0x20);
+				q_queries.pop();
+			}
+		}
```
Comment (Megamouse) [inline rpcs3/Emu/Io/Infinity.cpp:403]:
```suggestion
		else if (!q_queries.empty())
		{
			memcpy(buf, q_queries.front().data(), 0x20);
			q_queries.pop();
		}
```

File: rpcs3/Emu/Io/Infinity.cpp [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+	q_figure_added_removed_responses.push(figure_change_response);
+
+	return number;
+}
+
+usb_device_infinity::usb_device_infinity(const std::array<u8, 7>& location)
+	: usb_device_emulated(location)
+{
+	device = UsbDescriptorNode(USB_DESCRIPTOR_DEVICE, UsbDeviceDescriptor{0x200, 0x0, 0x0, 0x0, 0x20,
+														  0x0E6F, 0x0129, 0x200, 0x1, 0x2, 0x3, 0x1});
```
Comment (Megamouse) [inline rpcs3/Emu/Io/Infinity.cpp:363]:
```suggestion
	device = UsbDescriptorNode(USB_DESCRIPTOR_DEVICE, UsbDeviceDescriptor{0x200, 0x0, 0x0, 0x0, 0x20, 0x0E6F, 0x0129, 0x200, 0x1, 0x2, 0x3, 0x1});
```

File: rpcs3/Emu/Io/Infinity.cpp [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+	{
+		for (int x = 3; x >= 0; x--)
+		{
+			key[(3 - x) + (i * 4)] = output[x + (i * 4)];
+		}
+	}
+
+	aes_context aes;
+	aes_setkey_dec(&aes, key.data(), 128);
+	std::array<u8, 16> infinity_decrypted_block = {};
```
Comment (Megamouse) [inline rpcs3/Emu/Io/Infinity.cpp:332]:
```suggestion
	std::array<u8, 16> infinity_decrypted_block{};
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 5]
```diff
@@ -0,0 +1,821 @@
...
+	{0x3D09D6, {3, "Butch"}},
+	{0x3D09D7, {3, "Ramsey"}},
+	{0x3D09DC, {3, "Stars Over Sahara Square"}},
+	{0x3D09DD, {3, "Sahara Square Sands"}},
+	{0x3D09E0, {3, "Ghost Rider's Motorcycle"}},
+	{0x3D09E5, {3, "Quad Jumper"}}};
+
+u32 infinity_crc32(u16 init_value, const u8* buffer, u32 size)
+{
+	const u32 CRC32_TABLE[256] = {
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:361]:
```suggestion
	const std::array<u32, 256> CRC32_TABLE = {
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 6]
```diff
@@ -0,0 +1,821 @@
...
+
+			last_figure_path = QFileInfo(file_path).absolutePath() + "/";
+			accept();
+		});
+
+	connect(btn_cancel, &QAbstractButton::clicked, this, &QDialog::reject);
+
+	connect(co_compl, QOverload<const QString&>::of(&QCompleter::activated), [=](const QString& text)
+		{
+			combo_figlist->setCurrentText(text);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:543]: this is redundant

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 7]
```diff
@@ -0,0 +1,821 @@
...
+	fs::file inf_file(file_path.toStdString(), fs::read + fs::write + fs::create);
+	if (!inf_file)
+	{
+		return false;
+	}
+	// Create a 320 byte file with standard NFC read/write permissions
+	std::array<u8, 0x14 * 0x10> file_data{};
+	u32 first_block = 0x17878E;
+	u32 other_blocks = 0x778788;
+	for (s8 i = 2; i >= 0; i--)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:560]: why is this in reverse?

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 8]
```diff
@@ -0,0 +1,821 @@
...
+	{
+		file_data[0x38 - i] = u8((first_block >> i * 8) & 0xFF);
+	}
+	for (u32 index = 1; index < 0x05; index++)
+	{
+		for (s8 i = 2; i >= 0; i--)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:566]: why is this in reverse?

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 9]
```diff
@@ -0,0 +1,821 @@
...
+	}
+	// Create a 320 byte file with standard NFC read/write permissions
+	std::array<u8, 0x14 * 0x10> file_data{};
+	u32 first_block = 0x17878E;
+	u32 other_blocks = 0x778788;
+	for (s8 i = 2; i >= 0; i--)
+	{
+		file_data[0x38 - i] = u8((first_block >> i * 8) & 0xFF);
+	}
+	for (u32 index = 1; index < 0x05; index++)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:564]:
```suggestion
	for (u32 index = 1; index < 5; index++)
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 10]
```diff
@@ -0,0 +1,821 @@
...
+	{
+		for (s8 i = 2; i >= 0; i--)
+		{
+			file_data[((index * 0x40) + 0x38) - i] = u8((other_blocks >> i * 8) & 0xFF);
+		}
+	}
+	// Create the vector to calculate the SHA1 hash with
+	std::vector<u8> sha1_calc = {SHA1_CONSTANT.begin(), SHA1_CONSTANT.end() - 1};
+	// Generate random UID, used for AES encrypt/decrypt
+	std::array<u8, 16> uid_data = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x89, 0x44, 0x00, 0xC2};
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:574]:
```suggestion
	const std::array<u8, 16> uid_data = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x89, 0x44, 0x00, 0xC2};
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 11]
```diff
@@ -0,0 +1,821 @@
...
+	}
+	else if (series == 3)
+	{
+		figure_data[4] = 0x0F;
+		figure_data[5] = 0x08;
+		figure_data[6] = 0x1C;
+	}
+
+	u32 checksum = infinity_crc32(0, figure_data.data(), 12);
+	for (s8 i = 3; i >= 0; i--)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:609]: why is this in reverse?

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 12]
```diff
@@ -0,0 +1,821 @@
...
+	{
+		figure_data[15 - i] = u8((checksum >> i * 8) & 0xFF);
+	}
+
+	if (figure_data[1] == 0x00)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:614]:
```suggestion
	if (figure_data[1] == 0)
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 13]
```diff
@@ -0,0 +1,821 @@
...
+			key[(3 - x) + (i * 4)] = output[x + (i * 4)];
+		}
+	}
+
+	// Create AES Encrypt context based on AES key, use this to encrypt the character data and 4 blank
+	// blocks
+	aes_context aes;
+	aes_setkey_enc(&aes, key, 128);
+	std::array<u8, 16> encrypted_block = {};
+	std::array<u8, 16> encrypted_blank = {};
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:638]:
```suggestion
	std::array<u8, 16> encrypted_block{};
	std::array<u8, 16> encrypted_blank{};
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 14]
```diff
@@ -0,0 +1,821 @@
...
+#include <QFileDialog>
+#include <QVBoxLayout>
+#include <QMessageBox>
+#include <QComboBox>
+#include <QPushButton>
+#include <QStringList>
+#include <QCompleter>
+
+infinity_dialog* infinity_dialog::inst = nullptr;
+std::optional<u32> infinity_dialog::figure_slots[7];
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:24]: this is useless

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 15]
```diff
@@ -0,0 +1,821 @@
...
+	hbox_infinity->addWidget(clear_btn);
+	hbox_infinity->addWidget(create_btn);
+	hbox_infinity->addWidget(load_btn);
+
+	vbox_group->addLayout(hbox_infinity);
+}
+
+void infinity_dialog::clear_figure(u8 slot)
+{
+	if (auto slot_infos = figure_slots[slot])
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:762]:
```suggestion
	ensure(slot < figure_slots.size());

	if (figure_slots[slot])
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 16]
```diff
@@ -0,0 +1,821 @@
...
+	{
+		g_infinitybase.remove_figure(slot);
+		figure_slots[slot] = 0;
+		edit_figures[slot]->setText(tr("None"));
+	}
+}
+
+void infinity_dialog::create_figure(u8 slot)
+{
+	figure_creator_dialog create_dlg(this, slot);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:772]:
```suggestion
	ensure(slot < figure_slots.size());
	figure_creator_dialog create_dlg(this, slot);
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 17]
```diff
@@ -0,0 +1,821 @@
...
+	if (create_dlg.exec() == Accepted)
+	{
+		load_figure_path(slot, create_dlg.get_file_path());
+	}
+}
+
+void infinity_dialog::load_figure(u8 slot)
+{
+	const QString file_path = QFileDialog::getOpenFileName(this, tr("Select Infinity File"), last_figure_path, tr("Infinity Figure (*.bin);;"));
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:781]:
```suggestion
	ensure(slot < figure_slots.size());
	const QString file_path = QFileDialog::getOpenFileName(this, tr("Select Infinity File"), last_figure_path, tr("Infinity Figure (*.bin);;"));
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 18]
```diff
@@ -0,0 +1,821 @@
...
+#include <QVBoxLayout>
+#include <QMessageBox>
+#include <QComboBox>
+#include <QPushButton>
+#include <QStringList>
+#include <QCompleter>
+
+infinity_dialog* infinity_dialog::inst = nullptr;
+std::optional<u32> infinity_dialog::figure_slots[7];
+QString last_figure_path;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:25]: this should be a static member and not a global

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 19]
```diff
@@ -0,0 +1,821 @@
...
+	std::array<u8, 0x14 * 0x10> data;
+	if (inf_file.read(data.data(), data.size()) != data.size())
+	{
+		QMessageBox::warning(this, tr("Failed to read the figure file!"), tr("Failed to read the figure file(%1)!\nFile was too small.").arg(path), QMessageBox::Ok);
+		return;
+	}
+
+	clear_figure(slot);
+
+	u32 fignum = g_infinitybase.load_figure(data, std::move(inf_file), slot);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:810]:
```suggestion
	const u32 fignum = g_infinitybase.load_figure(data, std::move(inf_file), slot);
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 20]
```diff
@@ -0,0 +1,821 @@
...
+	auto name = list_figures.find(fignum);
+	if (name != list_figures.end())
+	{
+		edit_figures[slot]->setText(QString::fromStdString(name->second.second));
+	}
+    else
+    {
+        edit_figures[slot]->setText(tr("Unknown Figure"));
+    }
+    figure_slots[slot] = fignum;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:820]:
```suggestion
	else
	{
		edit_figures[slot]->setText(tr("Unknown Figure"));
	}
	figure_slots[slot] = fignum;
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 21]
```diff
@@ -0,0 +1,821 @@
...
+
+void infinity_dialog::add_figure_slot(QVBoxLayout* vbox_group, QString name, u8 slot)
+{
+	auto* hbox_infinity = new QHBoxLayout();
+
+	auto* label_figname = new QLabel(name);
+
+	auto* clear_btn = new QPushButton(tr("Clear"));
+	auto* create_btn = new QPushButton(tr("Create"));
+	auto* load_btn = new QPushButton(tr("Load"));
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:722]:
```suggestion
	QHBoxLayout* hbox_infinity = new QHBoxLayout();

	QLabel* label_figname = new QLabel(name);

	QPushButton* clear_btn = new QPushButton(tr("Clear"));
	QPushButton* create_btn = new QPushButton(tr("Create"));
	QPushButton* load_btn = new QPushButton(tr("Load"));
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 22]
```diff
@@ -0,0 +1,821 @@
...
+infinity_dialog* infinity_dialog::get_dlg(QWidget* parent)
+{
+	if (inst == nullptr)
+		inst = new infinity_dialog(parent);
+
+	return inst;
+}
+
+void infinity_dialog::add_figure_slot(QVBoxLayout* vbox_group, QString name, u8 slot)
+{
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:715]:
```suggestion
{
	ensure(slot < figure_slots.size());
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 23]
```diff
@@ -0,0 +1,821 @@
...
+	auto* hbox_infinity = new QHBoxLayout();
+
+	auto* label_figname = new QLabel(name);
+
+	auto* clear_btn = new QPushButton(tr("Clear"));
+	auto* create_btn = new QPushButton(tr("Create"));
+	auto* load_btn = new QPushButton(tr("Load"));
+	edit_figures[slot] = new QLineEdit();
+	edit_figures[slot]->setEnabled(false);
+	if (auto slot_infos = figure_slots[slot])
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:725]:
```suggestion
	if (figure_slots[slot])
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 24]
```diff
@@ -0,0 +1,821 @@
...
+	{
+		const auto found_figure = list_figures.find(figure_slots[slot].value());
+		if (found_figure != list_figures.end())
+		{
+			edit_figures[slot]->setText(QString::fromStdString(found_figure->second.second));
+		}
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:731]: what if it isn't found?

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 25]
```diff
@@ -0,0 +1,821 @@
...
+	u8 output[20];
+
+	sha1_starts(&ctx);
+	sha1_update(&ctx, sha1_calc.data(), sha1_calc.size());
+	sha1_finish(&ctx, output);
+
+	u8 key[0x10];
+	for (int i = 0; i < 4; i++)
+	{
+		for (int x = 3; x >= 0; x--)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:627]: why is this in reverse?

Comment (deReeperJosh) [reply rpcs3/rpcs3qt/infinity_dialog.cpp:627]: Just a quirk with Disney Infinity - their data is encrypted by an AES key which is the first 16 bytes of an SHA1 digest - but every 4 bytes needs to be reversed, will add a comment in here explaining this

Comment (Megamouse) [reply rpcs3/rpcs3qt/infinity_dialog.cpp:627]: I don't see how how this affects how you set the bytes.
a regular forward loop should be identical unless you set the same bytes twice

Comment (deReeperJosh) [reply rpcs3/rpcs3qt/infinity_dialog.cpp:627]: Sorry, have been confusing myself - updating the reversed ones now

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 26]
```diff
@@ -0,0 +1,831 @@
...
+	std::array<u8, 0x14 * 0x10> file_data{};
+	u32 first_block = 0x17878E;
+	u32 other_blocks = 0x778788;
+	for (u8 i = 0; i < 3; i++)
+	{
+		file_data[0x36 + i] = u8((first_block >> (2 - i) * 8) & 0xFF);
+	}
+	for (u32 index = 1; index < 5; index++)
+	{
+		for (u8 i = 0; i < 3; i--)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:565]:
```suggestion
		for (u8 i = 0; i < 3; i++)
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 27]
```diff
@@ -0,0 +1,831 @@
...
+#include <QVBoxLayout>
+#include <QMessageBox>
+#include <QComboBox>
+#include <QPushButton>
+#include <QStringList>
+#include <QCompleter>
+
+infinity_dialog* infinity_dialog::inst = nullptr;
+std::array<std::optional<u32>, 7> infinity_dialog::figure_slots = {};
+static QString s_last_figure_path;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:25]: I meant static member of the class

File: rpcs3/Emu/Io/Infinity.cpp [hunk 28]
```diff
@@ -0,0 +1,460 @@
...
+	u8 output[20];
+
+	sha1_starts(&ctx);
+	sha1_update(&ctx, sha1_calc.data(), sha1_calc.size());
+	sha1_finish(&ctx, output);
+
+	std::array<u8, 16> key = {};
+	for (int i = 0; i < 4; i++)
+	{
+		for (int x = 3; x >= 0; x--)
```
Comment (Megamouse) [inline rpcs3/Emu/Io/Infinity.cpp:308]:
```suggestion
		for (int x = 0; x < 4; x++)
```

File: rpcs3/Emu/Io/Infinity.cpp [hunk 29]
```diff
@@ -0,0 +1,460 @@
...
+	}
+
+	sha1_context ctx;
+	u8 output[20];
+
+	sha1_starts(&ctx);
+	sha1_update(&ctx, sha1_calc.data(), sha1_calc.size());
+	sha1_finish(&ctx, output);
+
+	std::array<u8, 16> key = {};
```
Comment (Megamouse) [inline rpcs3/Emu/Io/Infinity.cpp:305]:
```suggestion
	std::array<u8, 16> key{};
```

File: rpcs3/rpcs3qt/infinity_dialog.cpp (hunk 27 above)
Comment (deReeperJosh) [reply rpcs3/rpcs3qt/infinity_dialog.cpp:25]: It needs to be referenced in the create dialog and main dialog (2 separate classes), figured a global static string would suit here?

File: rpcs3/rpcs3qt/infinity_dialog.cpp [hunk 30]
```diff
@@ -0,0 +1,831 @@
...
+				QMessageBox::warning(this, tr("Error converting value"), tr("Figure number entered is invalid!"), QMessageBox::Ok);
+				return;
+			}
+			const u8 series = edit_series->text().toUShort(&ok_series);
+			if (!ok_series || series > 3 || series < 1)
+			{
+				QMessageBox::warning(this, tr("Error converting value"), tr("Series number entered is invalid!"), QMessageBox::Ok);
+				return;
+			}
+			QString predef_name = s_last_figure_path;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/infinity_dialog.cpp:513]: Redundant assignment
```suggestion
			QString predef_name;
```

File: rpcs3/Emu/Io/Infinity.cpp [hunk 31]
```diff
@@ -0,0 +1,460 @@
...
+
+void infinity_figure::save()
+{
+	if (!inf_file)
+	{
+		infinity_log.error("Tried to save infinity figure to file but no infinity figure is active!");
+		return;
+	}
+
+	{
```
Comment (Megamouse) [inline rpcs3/Emu/Io/Infinity.cpp:26]: unnecessary scope

File: rpcs3/Emu/Io/Infinity.cpp [hunk 32]
```diff
@@ -0,0 +1,460 @@
...
+	int x = 3;
+	for (u8 i = 0; i < figures.size(); i++)
+	{
+		u8 slot = i == 0 ? 0x10 : (i < 4) ? 0x20 :
+		                                    0x30;
+		if (figures[i].present)
+		{
+			reply_buf[x] = slot + figures[i].order_added;
+			reply_buf[x + 1] = 0x09;
+			x = x + 2;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/Infinity.cpp:166]:
```suggestion
			x += 2;
```

**Comment by Megamouse** [comment]
is this working now?

**Comment by deReeperJosh** [comment]
Works for me - have tried 1.0/2.0/3.0!

**Comment by shinra-electric** [comment]
Ignore that tag 👆🏻

**Comment by blan44** [comment]
The emulated base seems to be missing the feature to use multiple power disks at once.

**Comment by deReeperJosh** [comment]
> The emulated base seems to be missing the feature to use multiple power disks at once.

You can use 2 at once: 
Screenshot 2024-05-11 at 14 19 19

**Comment by blan44** [comment]
> > The emulated base seems to be missing the feature to use multiple power disks at once.
> 
> You can use 2 at once: Screenshot 2024-05-11 at 14 19 19

When I tried using that it would only recognise the first circular power disc and not the second one.
Also it should support up to three Hexagonal Power Discs instead of just one (so you can use vehicle power discs in Toy Box Speedway)

**Comment by deReeperJosh** [comment]
Thanks for the info, will have a look in to how to make that work. I wasn't aware that multiple circular power discs weren't working as expected!

**Comment by blan44** [comment]
> Thanks for the info, will have a look in to how to make that work. I wasn't aware that multiple circular power discs weren't working as expected!

Yeah I noticed something was off when I couldn't get the power disc combinations to activate.

**Comment by deReeperJosh** [comment]
I can't seem to build on my mac anymore, keep getting build errors, I assume because I am on the latest version of xcode - so won't be able to build/test any changes I make. I wrote the code for the Dolphin version as well, and power disc combinations work fine there, as well as I added the ability to have 3 discs in the hexagon slot (see here https://github.com/dolphin-emu/dolphin/pull/12771)

**Pull request merged**